# Importing Libraries

In [125]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing Data Sets

In [126]:
path = r'C:\Users\Aricia\MM-2020 Instacart Basket Analysis'

In [127]:
# Import Product Data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [128]:
# Import Orders Wrangled Data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Practice - df.prods

## Investigating using describe

In [129]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

## Data Test

In [130]:
# Create a dataframe
df_test = pd.DataFrame()

In [131]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [132]:
# Check for mixed types -  FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [133]:
# Changing Data type from numeric to string - fixing mixed data
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [134]:
# Finding Missing Values 
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [135]:
# Finding Missing Names in cells
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [136]:
# Checking names
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Replacing Missing Values with Mean (not to be used at the moment)
df['column with missings'].fillna(mean value, inplace=True)
df['column with missings'].fillna(median value, inplace=True)

In [137]:
# Counting number of rows and columns in a data frame
df_prods.shape

(49693, 5)

In [138]:
# Create a new data frame called clean without out null
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [139]:
# Check new count of rows and columns without null
df_prods_clean.shape

(49677, 5)

# Alternative Method to Clean Null Cells (inplace = True overwrites)
df_prods.dropna(subset = [‘product_name’], inplace = True)

## Duplicates

In [140]:
# Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [141]:
# Checking Duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [142]:
# Counting Rows and Columns in Dateframe 
df_prods_clean.shape

(49677, 5)

In [143]:
# Addressing Duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

Alternative method df_prods_clean.drop_duplicates()

In [145]:
# Check New Count of Rows and Columns without duplicates
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

# 4.5 Task - df. ords

## Task 4.5, 2 - Investigating using describe

In [146]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Answer: Order number max is 100 which can be a low value considering the size of data. All the other columns looks okay since they are within the expected.

## Task 4.5, item 2 - Check for mix data

In [147]:
# Check for mixed types -  FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Answer: There is no mixed-type data

## Task 4.5, items 5 and 6 - Missing Values

In [148]:
# Item 5 - Finding Missing Values 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [149]:
# Finding Missing Days in cells
df_nan_days = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [150]:
# Checking dates
df_nan_days

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

Answer item 5: I found 206,209 rows of missing data. When analysing some rows with the null cell, I noticed that all the order numbers = 1. This means that this might not be a dataset mistake, but it is actually related to first orders. As there were no orders before, the # of days prior order will not exist, which returns result null. Thefore, I would not substitute with mean or median, for example, as this would no be accurate and could change the data. 

In [151]:
# Replacing Missing Values with -1 (C:\Users\Aricia\AppData\Local\Temp\ipykernel_46376\159452663.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy).
df_ords['days_since_prior_order'].fillna(-1, inplace=True)

C:\Users\Aricia\AppData\Local\Temp\ipykernel_45768\516396208.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords['days_since_prior_order'].fillna(-1, inplace=True)


In [152]:
# Checking replaced values
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

Answer item 6: As mentioned on previous exercise, the Null cells were related to first orders, which logically did not have a past order. For that reason, I decided to replace values with a special value, that would not compromise the integrity of the data. For that reason, I decided to replace all null values for -1, as it clearly flags the rows as first orders, since days have to be = > 0, without interfering with the actual order already existent.

## Task 4.5, items 7 and 8 - Duplicates

In [153]:
# Finding Duplicates
df_dups_ords = df_ords[df_ords.duplicated()]

In [154]:
# Checking Duplicates
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Answer 7: There are no duplicates on the orders dataset.

# Exporting Data 

In [155]:
# Complete at the end to Prepared Data
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [156]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))